ui/library/src/utils/VisibilityDetector.ts

```diff
@@ -0,0 +1,69 @@
+export type SetEntriesCallback<TEntry> = (visible: Set<TEntry>) => void;
+export type OnVisibilityChangeCallback<TEntry> = (args: {
+  visibleElements: Element[];
+  hiddenElements: Element[];
+  lastEntries: Set<TEntry>;
+}) => Set<TEntry>;
+
+export default class VisibilityDetector<TEntry> {
```

Comment:
I would suggest using the name `VisibleEntriesDetector` for this class

Reply:
@yensung gotcha i will rename file and class name too so its more consistent.